### PR TITLE
Status bar element for Development mode

### PR DESCRIPTION
**Pull request by vitaliy-guliy**
Status bar element for development mode.

development-host

development-host-offline

https://www.youtube.com/watch?v=XRQ2f5bKRno

### CODE REVIEW

File: packages/plugin-ext/src/hosted/browser/hosted-plugin-informer.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    @inject(FrontendApplicationStateService)
+    protected readonly frontendApplicationStateService: FrontendApplicationStateService;
+
+    @postConstruct()
+    protected async init(): Promise<void> {
+        this.workspaceService.root.then(root => {
+            this.hostedPluginServer.getHostedPlugin().then(pluginMetadata => {
+                if (pluginMetadata) {
+                    this.entry = {
+                        text: `$(cube) ` + HostedPluginInformer.DEVELOPMENT_HOST,
```
Comment (olexii4) [inline packages/plugin-ext/src/hosted/browser/hosted-plugin-informer.ts:46]: Please, use 
```
`...${parameter}...`
```

File: packages/plugin-ext/src/hosted/browser/hosted-plugin-informer.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            this.entry.style = StatusBarEntryStyle.SUCCESS;
+        }
+
+        this.statusBar.setElement('development-host', this.entry);
+    }
+
+    private updateTitle(root: FileStat | undefined): void {
+        if (root) {
+            const uri = new URI(root.uri);
+            document.title = HostedPluginInformer.DEVELOPMENT_HOST + " - " + uri.displayName;
```
Comment (olexii4) [inline packages/plugin-ext/src/hosted/browser/hosted-plugin-informer.ts:76]: Please, use
```
`...${parameter}...`
```

**Review by akurinnoy — APPROVED** [review]

**Review by mmorhun — APPROVED** [review]

**Review by olexii4 — APPROVED** [review]

**Review by evidolob — APPROVED** [review]

**Comment by benoitf** [comment]
@vitaliy-guliy could you also make a PR to upstream for the "core" package ?
Goal is to have on this fork only stuff related to the plugin's model
so external changes should be merged as well upstream

**Review by benoitf — APPROVED** [review]

File: packages/plugin-ext/src/hosted/browser/hosted-plugin-informer.ts (hunk 1 above)
Comment (vitaliy-guliy) [reply packages/plugin-ext/src/hosted/browser/hosted-plugin-informer.ts:46]: Ok. Thank you.

File: packages/plugin-ext/src/hosted/browser/hosted-plugin-informer.ts (hunk 2 above)
Comment (vitaliy-guliy) [reply packages/plugin-ext/src/hosted/browser/hosted-plugin-informer.ts:76]: Ok, thanks.

**Comment by vitaliy-guliy** [comment]
PR to upstream with colored status bar elements
https://github.com/theia-ide/theia/pull/1840

File: packages/plugin-ext/src/hosted/browser/hosted-plugin-informer.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import { HostedPluginServer } from '../../common/plugin-protocol';
+import { ConnectionStatusService, ConnectionState } from '@theia/core/lib/browser/connection-status-service';
+import URI from '@theia/core/lib/common/uri';
+import { FileStat } from '@theia/filesystem/lib/common';
+import { FrontendApplicationStateService } from '@theia/core/lib/browser/frontend-application-state';
+
+@injectable()
+export class HostedPluginInformer {
+
+    public static readonly DEVELOPMENT_HOST = "Developement Host";
```
Comment (benoitf) [inline packages/plugin-ext/src/hosted/browser/hosted-plugin-informer.ts:21]: small typo on `Development`

File: packages/plugin-ext/src/hosted/browser/hosted-plugin-informer.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import { StatusBarAlignment, StatusBarEntryStyle, StatusBarEntry } from '@theia/core/lib/browser';
+import { WorkspaceService } from '@theia/workspace/lib/browser';
+import { HostedPluginServer } from '../../common/plugin-protocol';
+import { ConnectionStatusService, ConnectionState } from '@theia/core/lib/browser/connection-status-service';
+import URI from '@theia/core/lib/common/uri';
+import { FileStat } from '@theia/filesystem/lib/common';
+import { FrontendApplicationStateService } from '@theia/core/lib/browser/frontend-application-state';
+
+@injectable()
+export class HostedPluginInformer {
```
Comment (benoitf) [inline packages/plugin-ext/src/hosted/browser/hosted-plugin-informer.ts:19]: you could add some description of the class

File: packages/plugin-ext/src/main/browser/plugin-ext-frontend-module.ts [hunk 5]
```diff
@@ -15,25 +15,31 @@ import { HostedPluginManagerClient } from "./plugin-manager-client";
...
 
+    bind(HostedPluginInformer).toSelf().inSingletonScope();
+
     bind(PluginApiFrontendContribution).toSelf().inSingletonScope();
     bind(FrontendApplicationContribution).toDynamicValue(c => c.container.get(PluginApiFrontendContribution));
     bind(CommandContribution).toDynamicValue(c => c.container.get(PluginApiFrontendContribution));
 
     bind(FrontendApplicationContribution).toDynamicValue(ctx => ({
         onStart(app: FrontendApplication): MaybePromise<void> {
+            ctx.container.get(HostedPluginInformer);
```
Comment (benoitf) [inline packages/plugin-ext/src/main/browser/plugin-ext-frontend-module.ts:34]: IMHO it looks strange to get the instance from the container but not using it.
should you bind it along with
```bind(FrontendApplicationContribution).toDynamicValue(c => c.container.get(HostedPluginInformer));``` ? and have onStart method in your instance ?

Comment (akosyakov) [reply packages/plugin-ext/src/main/browser/plugin-ext-frontend-module.ts:34]: `HostedPluginInformer` should implement `FrontendApplicationContribution`, it is an entry point for frontend contributions

File: packages/plugin-ext/src/hosted/browser/hosted-plugin-informer.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            this.hostedPluginServer.getHostedPlugin().then(pluginMetadata => {
+                if (pluginMetadata) {
+                    this.entry = {
+                        text: `$(cube) ` + HostedPluginInformer.DEVELOPMENT_HOST,
+                        tooltip: `Hosted Plugin '` + pluginMetadata.model.name + `'`,
+                        alignment: StatusBarAlignment.LEFT,
+                        priority: 100
+                    };
+
+                    this.frontendApplicationStateService.reachedState('init').then(() => {
```
Comment (akosyakov) [inline packages/plugin-ext/src/hosted/browser/hosted-plugin-informer.ts:52]: I is an internal API for `FrontendApplication`, extension developers should implement lifecycle callbacks via `FrontendApplicationContribution`, in this case `FrontendApplicationContribution.onStart`

**Pull request closed without merging**